src/Microsoft.AspNet.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -0,0 +1,678 @@
+            }
+
+            if (offset < 0 || offset >= mediaType.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (length != null && offset + length > mediaType.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            _mediaType = mediaType;
+            _offset = offset;
+            _length = length;
+            _parameterEnumerator = default(Enumerator);
+            Type = new StringSegment();
+            SubType = new StringSegment();
+
+            var enumerator = GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                return;
+            }
+            Type = enumerator.Current.Value;
+
+            if (!enumerator.MoveNext())
+            {
+                return;
+            }
+            SubType = enumerator.Current.Value;
+
+            _parameterEnumerator = enumerator;
+        }
+
+        /// <summary>
+        /// Gets the type of the <see cref="MediaType"/>.
+        /// </summary>
+        public StringSegment Type { get; }
+
+        /// <summary>
+        /// Gets whether this <see cref="MediaType"/> matches all types.
+        /// </summary>
+        public bool MatchesAllTypes => Type.Equals("*", StringComparison.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Gets the subtype of the <see cref="MediaType"/>.
+        /// </summary>
+        public StringSegment SubType { get; private set; }
+
+        /// <summary>
+        /// Gets whether this <see cref="MediaType"/> matches all subtypes.
+        /// </summary>
+        public bool MatchesAllSubTypes => SubType.Equals("*", StringComparison.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Gets whether this <see cref="MediaType"/> has a valid value.
+        /// </summary>
+        public bool HasValue => !_parameterEnumerator.ParsingFailed;
+
+        /// <summary>
+        /// Gets the <see cref="System.Text.Encoding"/> of the <see cref="MediaType"/> if it has one.
+        /// </summary>
+        public Encoding Encoding => GetEncodingFromCharset(GetParameter("charset"));
+
+        /// <summary>
+        /// Gets the charset parameter of the <see cref="MediaType"/> if it has one.
+        /// </summary>
+        public StringSegment Charset => GetParameter("charset");
+
+        /// <summary>
+        /// Determines whether the current <see cref="MediaType"/> is a subset of the <paramref name="set"/> <see cref="MediaType"/>.
+        /// </summary>
+        /// <param name="set">The set <see cref="MediaType"/>.</param>
+        /// <returns>
+        /// <code>true</code> if this <see cref="MediaType"/> is a subset of <paramref name="set"/>; otherwise<code>false</code>.
+        /// </returns>
+        public bool IsSubsetOf(MediaType set)
+        {
+            return HasValue && set.HasValue &&
```

Comment:
@rynowak I think that if we allow invalid media types through the system is better that we are just resilient about them, if we were to throw an exception for an invalid media type wouldn't it be better to have it throw it as soon as possible rather than later in the execution?

The implication of returning false on IsSubsetOf with an invalid media type is that you'll get a 406 (haven't verified this, but is my expectation) which I think is totally valid and easy enough for a user/client to discover. The scenarios I can foresee for this are:

Client sends bad media type in accept header. I would expect here Content-Negotiation to fail or just return whatever predetermined format the server has configured, (like json). It's up to the server.

Server sets the content-type to be an invalid media type. If we throw, then we would need to know what to do with the exception or the user would have to handle it, if we don't you would likely get a 406 with the appropriate error message or you would get a predetermined format like json.